src/common/tse.c

```diff
@@ -505,22 +507,29 @@ tse_task_complete_callback(tse_task_t *task)
 }
 
 /*
- * Process the task in the init list of the scheduler. This executes all the
- * body function of all tasks with no dependencies in the scheduler's init
- * list.
+ * Process the init and sleeping lists of the scheduler. This first moves all
+ * tasks who shall wake up now from the sleeping list to the tail of the init
+ * list, and then executes all the body functions of all tasks with no
+ * dependencies in the scheduler's init list.
  */
 static int
 tse_sched_process_init(struct tse_sched_private *dsp)
 {
 	struct tse_task_private		*dtp;
 	struct tse_task_private		*tmp;
 	d_list_t			list;
+	uint64_t			now = daos_getutime();
 	int				processed = 0;
 
 	D_INIT_LIST_HEAD(&list);
 	D_MUTEX_LOCK(&dsp->dsp_lock);
-	d_list_for_each_entry_safe(dtp, tmp, &dsp->dsp_init_list,
-				      dtp_list) {
+	d_list_for_each_entry_safe(dtp, tmp, &dsp->dsp_sleeping_list,
+				   dtp_list) {
+		if (dtp->dtp_wakeup_time > now)
```

Comment:
I didn't find where you set the wakeup_time for a task.

Reply:
Yes, haven't added that. Will be in the new set_delay function.